torch/csrc/distributed/c10d/init.cpp

```diff
@@ -1018,6 +1068,16 @@ that adds a prefix to each key inserted to the store.
               [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
                  const std::vector<at::Tensor>& tensors,
                  const ::c10d::AllreduceOptions& opts) {
+                if (opts.reduceOp.op_ == ::c10d::ReduceOp::PREMUL_SUM) {
+#if defined(ENABLE_NCCL_PREMUL_SUM_SUPPORT)
+                  if (self->getBackendName() == "nccl") {
+                    std::vector<at::Tensor> non_const_tensors = tensors;
+                    return self->allreduce(non_const_tensors, opts);
```

Comment:
Essentially, you are bypassing the dispatcher by calling `self->allreduce` here. It would lead to an escape of traceability by the PyTorch compiler.
Temporarily it might be okay, but in long term we are going to require all ops go through the dispatcher.

Reply:
Regarding your question of how the dispatcher can recognize custom data types, I wonder if this reference code could help:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/Ops.cpp#L142-L145

Reply:
yeah I know this is dodging the effort of making c10d ops traceable. 
this is my "workaround" to keep "Ops" clean due to my lack of knowledge on TorchScript.
let's access this in a follow-up PR